pycoal/mineral.py

```diff
@@ -2,132 +2,217 @@
+        self.wavelength_max = 2.5082
+
+        # define number of layers in normalized pixel
+        self.num_layers = 128
+
+    def normalize(self, image, wavelengthList=None):
+        """
+        Return a copy of an MxNxB image (or 1x1xB pixel) with constrained wavelength range and normalized spectra.
+        Args:
+           image (numpy.ndarray): the image or pixel to normalize
+           wavelengthList (list) [default None]: the list of wavelengths for the pixel
+        Returns:
+           normalizedImage (numpy.ndarray): the normalized image or pixel
+        """
+
+        # constrain wavelength range
+        if wavelengthList is None:
+            wavelengthList = image.metadata.get('wavelength')
+
+        # get the indices of the minimum and maximum element
+        i = self.__indexOfGreaterThan(wavelengthList, self.wavelength_min)
+        j = self.__indexOfGreaterThan(wavelengthList, self.wavelength_max)
+
+        # constrain wavelength range
+        # slice the array from MxNxB to MxNx(j-i) for indices i,j
+        image = image[:,:,i:j]
+
+        # normalize spectra
+        principalComponents = spectral.algorithms.principal_components(image)
+        principalComponents.reduce(self.num_layers)
+        normalizedImage = principalComponents.transform(image)
+
+        # return normalized image
+        return normalizedImage
+
+    def __indexOfGreaterThan(self, elements, value):
+        """
+        Return the index of the first element greater than the value.
+        Args:
+           elements (list): list of elements
+           value (float): the value to compare
+        Returns:
+           index (int): the index of the first element greater than the value
+        """
+
+        for index,element in enumerate(elements):
+            if element > value:
+                return index
+
+    def trainClassifier(self, libraryFileName, classifierType='gaussian'):
+        """
+        This takes the file name of the library to be used in the training and returns a
+        classifier
+        Args:
+           libraryFileName (str): the name of the training library
+           classifierType (str) [default 'gaussian']: the type of classifier.
+                                                      Currently, can be one of:
+                                                      perceptron, gaussian, or
+                                                      mahalanobisdistance.
+        Returns:
+           classifier (PerceptronClassifier): trained classifier
+        """
+
+        # open the digital spectral library
+        # we may need to load it as an image instead
+        library = spectral.io.envi.open(libraryFileName)
+
+        # get dimensions of library to create mask for training class
+        dimensions = library.spectra.shape
+
+        # initialize the type of classifier
+        if classifierType == 'perceptron':
+            classifier = spectral.classifiers.PerceptronClassifier(???)
```

Comment:
Yep, this is all that's left for making this at correct syntax wise. I'm aiming to have this done by our meeting. Thanks!